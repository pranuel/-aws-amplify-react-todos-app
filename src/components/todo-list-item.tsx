import React, { ChangeEvent } from "react";
import { isEnterKey } from "../helpers/keyboard.helper";
import { Todo } from "../todos/todo.model";

export interface TodoListItemProps {
  todo: Todo;
  editTodo?: Todo;
  onEditTodo: (todo: Todo) => void;
  onTodoIsDoneToggled: (todo: Todo, isDone: boolean) => void;
  onSaveTodo: (todo: Todo) => void;
  onDeleteTodo: (todo: Todo) => void;
  onEditingTodoDescriptionChanged: (
    editTodo: Todo,
    description: string,
  ) => void;
}

export class TodoListItem extends React.Component<TodoListItemProps, {}> {
  public render() {
    const { todo, editTodo } = this.props;
    return (
      <li
        className={this.getTodoListItemClassName(todo, editTodo)}
        onDoubleClick={this.onDoubleClick}
      >
        <div className="view">
          <input
            className="toggle"
            type="checkbox"
            data-testid="is-done-toggle"
            checked={todo.isDone}
            onChange={this.onTodoIsDoneToggled}
          />
          <label>{todo.description}</label>
          <button
            className="destroy"
            data-testid="delete-button"
            onClick={this.onDeleteTodoClicked}
          />
        </div>
        <input
          className="edit"
          value={!!editTodo ? editTodo.description : ""}
          data-testid="edit-input"
          onKeyDown={this.onEditInputKeyDown}
          onBlur={this.onEditInputBlur}
          onChange={this.onEditInputChanged}
        />
      </li>
    );
  }

  private onEditInputChanged = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (!this.props.editTodo) {
      return;
    }
    this.props.onEditingTodoDescriptionChanged(
      this.props.editTodo,
      event.target.value,
    );
  }

  private onDeleteTodoClicked = () => {
    this.props.onDeleteTodo(this.props.todo);
  }

  private onEditInputBlur = () => {
    if (!this.props.editTodo) {
      return;
    }
    this.props.onSaveTodo(this.props.editTodo);
  }

  private onTodoIsDoneToggled = (event: ChangeEvent<HTMLInputElement>) => {
    this.props.onTodoIsDoneToggled(this.props.todo, event.target.checked);
  }

  private onEditInputKeyDown = (
    event: React.KeyboardEvent<HTMLInputElement>,
  ) => {
    if (!isEnterKey(event.key) || !this.props.editTodo) {
      return;
    }
    this.props.onSaveTodo(this.props.editTodo);
  }

  private onDoubleClick = () => {
    this.props.onEditTodo(this.props.todo);
  }

  private getTodoListItemClassName(todo: Todo, editTodo?: Todo): string {
    const classNames = [];
    if (todo.isDone) {
      classNames.push("completed");
    }
    if (!!editTodo && editTodo.id === todo.id) {
      classNames.push("editing");
    }
    return classNames.join(" ");
  }
}
