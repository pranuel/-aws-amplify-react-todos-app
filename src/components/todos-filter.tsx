import React from "react";
import { TodoFilter } from "../todos/todo-filter.model";

interface TodosFilterProps {
  todoFilter: TodoFilter;
  onShowAllTodosClicked: () => void;
  onShowActiveTodosClicked: () => void;
  onShowCompletedTodosClicked: () => void;
}

export class TodosFilter extends React.Component<TodosFilterProps, {}> {
  public render() {
    return (
      <ul className="filters">
        <li>
          <a
            href="#/"
            onClick={this.props.onShowAllTodosClicked}
            className={this.getTodosFilterClassName(TodoFilter.All)}
          >
            All
          </a>
        </li>
        <span />
        <li>
          <a
            href="#/"
            onClick={this.props.onShowActiveTodosClicked}
            className={this.getTodosFilterClassName(TodoFilter.Active)}
          >
            Active
          </a>
        </li>
        <span />
        <li>
          <a
            href="#/"
            onClick={this.props.onShowCompletedTodosClicked}
            className={this.getTodosFilterClassName(TodoFilter.Completed)}
          >
            Completed
          </a>
        </li>
      </ul>
    );
  }

  private getTodosFilterClassName = (elementTodoFilter: TodoFilter) => {
    return this.props.todoFilter === elementTodoFilter ? "selected" : undefined;
  }
}
