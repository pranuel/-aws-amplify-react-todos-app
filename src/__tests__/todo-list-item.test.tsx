import { fireEvent, render } from "@testing-library/react";
import React from "react";
import renderer from "react-test-renderer";
import { TodoListItem, TodoListItemProps } from "../components/todo-list-item";
import { Todo } from "../todos/todo.model";

describe("TodoListItem", () => {
  let testProps: TodoListItemProps;

  beforeEach(() => {
    testProps = {
      todo: new Todo(),
      onEditTodo: jest.fn(),
      onTodoIsDoneToggled: jest.fn(),
      onSaveTodo: jest.fn(),
      onDeleteTodo: jest.fn(),
      onEditingTodoDescriptionChanged: jest.fn(),
    };
  });

  it("should mark an item as completed", () => {
    // arrange
    testProps.todo.isDone = true;
    // act
    const tree = renderer.create(<TodoListItem {...testProps} />).toJSON();
    // assert
    expect(tree).toMatchSnapshot();
  });

  it("should mark an item as editing", () => {
    // arrange
    testProps.editTodo = testProps.todo;
    // act
    const tree = renderer.create(<TodoListItem {...testProps} />).toJSON();
    // assert
    expect(tree).toMatchSnapshot();
  });

  it("should notify about a deleted todo when delete button clicked", () => {
    // arrange
    const onDeleteTodoSpy = spyOn(testProps, "onDeleteTodo");
    const { getByTestId } = render(<TodoListItem {...testProps} />);
    // act
    fireEvent.click(getByTestId("delete-button"));
    // assert
    expect(onDeleteTodoSpy).toHaveBeenCalledWith(testProps.todo);
  });

  it("should notify about a todo is done when checkbox was toggled", () => {
    // arrange
    const onTodoIsDoneToggledSpy = spyOn(testProps, "onTodoIsDoneToggled");
    const { getByTestId } = render(<TodoListItem {...testProps} />);
    // act
    fireEvent.click(getByTestId("is-done-toggle"));
    // assert
    expect(onTodoIsDoneToggledSpy).toHaveBeenCalledWith(testProps.todo, true);
  });

  it("should notify about a todo being edited when entry double-clicked", () => {
    // arrange
    const onEditTodoSpy = spyOn(testProps, "onEditTodo");
    testProps.todo.description = "foo todo";
    const { getByText } = render(<TodoListItem {...testProps} />);
    // act
    fireEvent.doubleClick(getByText("foo todo"));
    // assert
    expect(onEditTodoSpy).toHaveBeenCalledWith(testProps.todo);
  });
});
