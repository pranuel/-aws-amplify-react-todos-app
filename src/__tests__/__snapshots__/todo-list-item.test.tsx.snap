// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListItem should mark an item as completed 1`] = `
<li
  className="completed"
  onDoubleClick={[Function]}
>
  <div
    className="view"
  >
    <input
      checked={true}
      className="toggle"
      data-testid="is-done-toggle"
      onChange={[Function]}
      type="checkbox"
    />
    <label>
      
    </label>
    <button
      className="destroy"
      data-testid="delete-button"
      onClick={[Function]}
    />
  </div>
  <input
    className="edit"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    value=""
  />
</li>
`;

exports[`TodoListItem should mark an item as editing 1`] = `
<li
  className="editing"
  onDoubleClick={[Function]}
>
  <div
    className="view"
  >
    <input
      checked={false}
      className="toggle"
      data-testid="is-done-toggle"
      onChange={[Function]}
      type="checkbox"
    />
    <label>
      
    </label>
    <button
      className="destroy"
      data-testid="delete-button"
      onClick={[Function]}
    />
  </div>
  <input
    className="edit"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    value=""
  />
</li>
`;
